diff -rNu linuxptp-1.8.orig/phc2sys.c linuxptp-1.8/phc2sys.c
--- linuxptp-1.8.orig/phc2sys.c	2016-11-04 19:38:56.000000000 +0100
+++ linuxptp-1.8/phc2sys.c	2017-08-28 00:12:11.017587019 +0200
@@ -418,6 +418,8 @@
 {
 	int direction = node->forced_sync_offset;
 
+	if (direction > 0)
+		return (int64_t)0;
 	if (!direction)
 		direction = dst->is_utc - node->master->is_utc;
 	return (int64_t)dst->sync_offset * NS_PER_SEC * direction;
@@ -1223,7 +1225,7 @@
 	char *src_name = NULL, *dst_name = NULL;
 	struct clock *src, *dst;
 	struct config *cfg;
-	int autocfg = 0, rt = 0;
+	int autocfg = 0, rt = 0, nooff = 0;
 	int c, domain_number = 0, pps_fd = -1;
 	int r = -1, wait_sync = 0;
 	int print_level = LOG_INFO, use_syslog = 1, verbose = 0;
@@ -1251,11 +1253,15 @@
 	progname = strrchr(argv[0], '/');
 	progname = progname ? 1+progname : argv[0];
 	while (EOF != (c = getopt(argc, argv,
-				  "arc:d:s:E:P:I:S:F:R:N:O:L:M:i:u:wn:xz:l:mqvh"))) {
+				  "aArc:d:s:E:P:I:S:F:R:N:O:L:M:i:u:wn:xz:l:mqvh"))) {
 		switch (c) {
 		case 'a':
 			autocfg = 1;
 			break;
+		case 'A':
+			autocfg = 1;
+			nooff = 1;
+			break;
 		case 'r':
 			rt++;
 			break;
@@ -1414,6 +1420,8 @@
 			goto end;
 		if (auto_init_ports(&node, rt) < 0)
 			goto end;
+		if (nooff)
+			node.forced_sync_offset = 1;
 		r = do_loop(&node, 1);
 		goto end;
 	}
